src/upgrade.py

```diff
@@ -0,0 +1,219 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Dependency model for MySQL."""
+
+import json
+import logging
+from typing import TYPE_CHECKING
+
+from charms.data_platform_libs.v0.upgrade import (
+    ClusterNotReadyError,
+    DataUpgrade,
+    DependencyModel,
+    UpgradeFinishedEvent,
+    VersionError,
+)
+from charms.mysql.v0.mysql import MySQLSetClusterPrimaryError
+from ops.charm import ActionEvent, UpgradeCharmEvent
+from ops.model import BlockedStatus
+from pydantic import BaseModel
+from tenacity import RetryError, Retrying
+from tenacity.stop import stop_after_attempt
+from tenacity.wait import wait_fixed
+from typing_extensions import override
+
+from k8s_helpers import KubernetesClientError
+
+if TYPE_CHECKING:
+    from charm import MySQLOperatorCharm
+
+logger = logging.getLogger(__name__)
+
+
+class MySQLK8sDependenciesModel(BaseModel):
+    """MySQL dependencies model."""
+
+    charm: DependencyModel
+    rock: DependencyModel
+
+
+def get_mysql_k8s_dependencies_model() -> MySQLK8sDependenciesModel:
+    """Return the MySQL dependencies model."""
+    with open("src/dependency.json") as dependency_file:
+        _deps = json.load(dependency_file)
+    return MySQLK8sDependenciesModel(**_deps)
+
+
+class MySQLK8sUpgrade(DataUpgrade):
+    """MySQL upgrade class."""
+
+    def __init__(self, charm: "MySQLOperatorCharm", **kwargs) -> None:
+        """Initialize the class."""
+        super().__init__(charm, **kwargs)
+        self.charm = charm
+
+        self.framework.observe(
+            getattr(self.charm.on, "resume_upgrade_action"), self._on_resume_upgrade
+        )
+
+        self.framework.observe(getattr(self.charm.on, "mysql_pebble_ready"), self._on_pebble_ready)
+
+    @property
+    def highest_ordinal(self) -> int:
+        """Return the max ordinal."""
+        return len(self.peer_relation.units)
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        """Run pre-upgrade checks."""
+        fail_message = "Pre-upgrade check failed. Cannot upgrade."
+
```

Comment:
we need to run ` util.checkForServerUpgrade()` command of mysql-shell here
https://dev.mysql.com/doc/mysql-shell/8.0/en/mysql-shell-utilities-upgrade.html

Reply:
Considering that at this point we don't know to which version we will be upgrading to, what's the point of doing so here?

Reply:
At least to the current version for k8s and rerun it when they know the new version (e.g. after the first member updated on k8s)